### PR TITLE
Add custom_fields to vm resource and update documentation

### PR DESCRIPTION
**Pull request by Bobonium**

### CODE REVIEW

**Comment by smutel** [comment]
I tested your change using the code you put in the documentation.
I got this error message:
```
Error: [POST /virtualization/virtual-machines/][400] virtualization_virtual-machines_create default  map[custom_fields:[Invalid custom field for virtualization | virtual machine objects: foo]]
```

**Comment by Bobonium** [comment]
Custom fields have to be predefined through netbox before they can be used for virtual machine definitions. Did you create  a custom field for the key `foo` before trying the example from the documentation?

I tested this at work without any problems.

**Comment by smutel** [comment]
Yes ... I forgot this step. Now it's working ...

However I test this:
- I added a new custom fields for a VM
- I created a VM with this custom fields
- I removed the custom fields in tf and apply again
- No change has been applied

File: netbox/resource_netbox_virtualization_vm.go [hunk 1]
```diff
@@ -90,6 +90,11 @@ func resourceNetboxVirtualizationVM() *schema.Resource {
 				Optional: true,
 				Default:  0,
 			},
+            "custom_fields": &schema.Schema{
+                Type:     schema.TypeMap,
+                Optional: true,
+                Elem:     &schema.Schema{Type: schema.TypeString},
+            },
```
Comment (smutel) [inline netbox/resource_netbox_virtualization_vm.go:97]:
```suggestion
                        "custom_fields": &schema.Schema{
                                Type:     schema.TypeMap,
                                Optional: true,
                                Elem:     &schema.Schema{Type: schema.TypeString},
                        },
```

File: netbox/resource_netbox_virtualization_vm.go [hunk 2]
```diff
@@ -110,6 +115,7 @@ func resourceNetboxVirtualizationVMCreate(d *schema.ResourceData,
 	tags := d.Get("tag").(*schema.Set).List()
 	tenantID := int64(d.Get("tenant_id").(int))
 	vcpus := int64(d.Get("vcpus").(int))
+    customFields := d.Get("custom_fields").(map[string]interface{})
```
Comment (smutel) [inline netbox/resource_netbox_virtualization_vm.go:118]:
```suggestion
        customFields := d.Get("custom_fields").(map[string]interface{})
```

File: netbox/resource_netbox_virtualization_vm.go [hunk 3]
```diff
@@ -247,6 +256,10 @@ func resourceNetboxVirtualizationVMRead(d *schema.ResourceData,
 				return err
 			}
 
+		    if err = d.Set("custom_fields", resource.CustomFields); err != nil {
+		        return err
+		    }
```
Comment (smutel) [inline netbox/resource_netbox_virtualization_vm.go:261]:
```suggestion
		        if err = d.Set("custom_fields", resource.CustomFields); err != nil {
		                return err
		        }
```

File: netbox/resource_netbox_virtualization_vm.go [hunk 4]
```diff
@@ -314,6 +327,11 @@ func resourceNetboxVirtualizationVMUpdate(d *schema.ResourceData,
 		params.Vcpus = &vcpus
 	}
 
+	if d.HasChange("custom_fields") {
+        customFields := d.Get("custom_fields").(map[string]interface{})
+        params.CustomFields = &customFields
```
Comment (smutel) [inline netbox/resource_netbox_virtualization_vm.go:332]:
```suggestion
                customFields := d.Get("custom_fields").(map[string]interface{})
                params.CustomFields = &customFields
```

**Review by smutel — CHANGES_REQUESTED** [review]

**Comment by Bobonium** [comment]
weird, I tested updates as well and they were correctly displayed for me. I'll have another look at it tomorrow and see if I can find the problem.

**Comment by smutel** [comment]
> weird, I tested updates as well and they were correctly displayed for me. I'll have another look at it tomorrow and see if I can find the problem.

I tried to implement this feature previously and it's not so simple.
Do you manage here only text custom fields or you tried also other type like multiple choices, boolean, ... ?

**Comment by Bobonium** [comment]
that's a good point actually. In our case we only have text fields. To be honest I didn't even give it that much thought, yet. I merely added what was necessary for us to be able to make proper use of your provider. I can definitely have another look to see if I can find a solution to support all custom fields properly.

**Comment by Bobonium** [comment]
Unfortunately it wasn't as easy and as straightforward as I thought at first. Main reason being, that there's a technical limitation within Terraform that makes it impossible to have different data types in a map.

The recommended way is to convert everything to a string within terraform, even if it's not perfect it solves the problem.
Funnily enough the code was working already for creating and updating resources with all types of custom_fields if you supplied the values correctly (i.e. for dropdowns the ID is required).

```
			customFields := make(map[string]string)
			switch t := resource.CustomFields.(type) {
			case map[string]interface{}:
				for key, value := range t {
					var strValue string
					if value != nil {
						switch v := value.(type) {
						default:
							strValue = fmt.Sprintf("%v", v)
						case map[string]interface{}:
							strValue = fmt.Sprintf("%v", v["value"])
						}
					}
					customFields[key] = strValue

				}
			}
```

The `custom_fields` are always of type `map[string]interface{}` but the `interface` implementation differs based on the custom_fields. For literal strings, it's straightforward, but dropdowns are a bit more complex as the interface in this case is anohter `map[string]interface{}` This second map then contains 2 hardcoded keys ( `label`  and `value` ), where the `value` key always holds the data that we need to supply to the API


Here's a short terraform example: 
```
  custom_fields = {
    drop_down_value = "7"
    date_value      = "2020-11-30"
    string_value    = "hello world"
  }
```

**Comment by smutel** [comment]
Thank you for your updates on this PR.

I tested it a little bit this PR. The main issue is that when we remove the terraform code related to custom fields, the custom fields are not removed in Netbox.

Creation of text: OK
Creation of date: OK
Creation of integer: OK
Creation of boolean: NOT OK
Creation of url: OK (it's like string in fact, netbox is just displaying as a link in the web UI)
Creation of selection: OK but I am not fan of choosing the item with his position. I think we can do it better using the endpoint /api/extras/_custom_field_choices/. The issue is to identify that it's a selection.

Previously I tried to manage custom_fields like this:
```hcl
custom_field = {
    name = "drop_down_value"
    type   = "selection"
    value    = "test1"
}
custom_field = {
    name = "string"
    type   = "string"
    value    = "test2"
}
```
In this case we need a function to convert this terraform code in map[string]interface{}.
If you have time to work on this, it will be great. If not I will do it but I don't know when.

**Comment by Bobonium** [comment]
I had another look based on what you posted, and I think I can fix the situation with the code not being deleted. The problem seems to be, that terraform defaults these fields to `null`, but the Netbox API wants to have an empty String.
Boolean should also easily be fixable.

I'm not an expert in building providers, but have used terraform for a long time, and as far as I know it seems (without really looking at the code) that your syntax would internally be handled as a list. Most resources in providers I used, that have this kinda Syntax also have a lot of unnecessary changes just by removing one entry or by adding one value in the middle of things

Also if we have a look at other providers, i.e. the Amazon one. Then we can see that their way of going is not to automatically find ID's magically by doing name based lookups, they simply supply an additional resource to do a name lookup and you're supposed to use the output of that resource. This would also help to now blow up the VM code unnecessarily here.

**Comment by Bobonium** [comment]
Since booleans were still not working, I finally setup my own custom netbox instance for testing, to verify that this time all fields work. Feel free to double check.

I also added some comments to the code parts that are not really straight forward, to explain why we need them in the first place.

- all fields should now work
- fields removed from terraform will be removed from netbox
- deleting the block will remove all values from netbox
- multiple applies display the changes only once, afterwards the state has no more changes
- all values need to be supplied as strings
  - terraform limitation, as it does not yet support maps with multiple data types in them
- selection fields are chosen by ID instead of name
  - this I kept intentionally for now, as I think it makes more sense to have a separate data object, with which we can lookup the ID by name and then supply the ID based on the output dynamically

@smutel please have another look and let me know what you think

**Comment by smutel** [comment]
Thank you, I will test this new part this week.

Agree to have another resource to find the id of selection.

Do you think it's hard to implement boolean?

**Comment by Bobonium** [comment]
I actually already added boolean support with commit ID `0f33c42178adba7e6827b9c3a9cfcacb04948730`.
The problem was that it's literally the only custom_field value in the API, that is not expected as a string, but as a literal `0` or `1`.

File: docs/resources/virtualization_vm.md [hunk 5]
```diff
@@ -16,6 +16,10 @@ resource "netbox_virtualization_vm" "vm_test" {
     name = "tag1"
     slug = "tag1"
   }
+
+  custom_fields = {
+    foo = "bar"
+  }
```
Comment (smutel) [inline docs/resources/virtualization_vm.md:22]:
```suggestion
  custom_fields = {
    cf_boolean = "true"
    cf_date = "2020-12-25"
    cf_integer = "10"
    cf_selection = "1"
    cf_text = "Some text"
    cf_url = "https://github.com"
  }
```

File: docs/resources/virtualization_vm.md [hunk 6]
```diff
@@ -36,6 +40,7 @@ The following arguments are supported:
 The ``tag`` block supports:
 * ``name`` - (Required) Name of the existing tag to associate with this resource.
 * ``slug`` - (Required) Slug of the existing tag to associate with this resource.
+* ``custom_fields`` - (Optional) Custom Field Keys and Values for this object
```
Comment (smutel) [inline docs/resources/virtualization_vm.md:43]:
```suggestion
* ``custom_fields`` - (Optional) Custom Field Keys and Values for this object
** For boolean, use the string value "true" or "false"
** For data, use the string format "YYYY-MM-DD"
** For integer, use the value between double quote "10"
** For selection, use the level id
** For text, use the string value
** For URL, use the URL as string
```

File: netbox/resource_netbox_virtualization_vm.go [hunk 7]
```diff
@@ -314,6 +370,29 @@ func resourceNetboxVirtualizationVMUpdate(d *schema.ResourceData,
 		params.Vcpus = &vcpus
 	}
 
+	//
+	if d.HasChange("custom_fields") {
+		stateCustomFields, resourceCustomFields := d.GetChange("custom_fields")
+		customFields := make(map[string]interface{})
+		// netbox needs explicit empty string to remove old values
+		// first we fill all existing fields from the state with an empty string
+		for key, _ := range stateCustomFields.(map[string]interface{}) {
```
Comment (smutel) [inline netbox/resource_netbox_virtualization_vm.go:379]:
```suggestion
		for key := range stateCustomFields.(map[string]interface{}) {
```

**Review by smutel — CHANGES_REQUESTED** [review]
Thank you for this new commits. One last issue to resolve:
* When the boolean custom field are removed from terraform code, it's not removed from netbox.

Some improvements to do:
* Move some part of your code in [util.go](https://github.com/smutel/terraform-provider-netbox/blob/master/netbox/util.go) because we will need it for custom fields of other netbox resources.
* Update resource vm in example code https://github.com/smutel/terraform-provider-netbox/blob/a9fbe374499e2cd39dcc3dbff4ab4a6c6d974270/examples/main.tf#L106 to add this:
```hcl
custom_fields = {
    cf_boolean = "true"
    cf_date = "2020-12-25"
    cf_integer = "10"
    cf_selection = "1"
    cf_text = "Some text"
    cf_url = "https://github.com"
  }
```

**Comment by Bobonium** [comment]
Thank you for your feedback, I'll have another look at the boolean deletion problem. And while I'm add it I'll do the adjustments to move some re-usable stuff into util

**Comment by Bobonium** [comment]
Now all fields can be resetted as well. Setting `nil` explicitly instead of an empty string seems to work for all types of fields.

**Comment by Bobonium** [comment]
@smutel we're nearing the end. Boolean deletion now works and I've moved the special handlings into the util.go. Let me know if you need me to change anything else to get this merged.

**Review by smutel — APPROVED** [review]

**Comment by smutel** [comment]
Thanks @Bobonium

**Pull request merged**